### PR TITLE
Add coins to teams and user add/remove route

### PR DESCRIPTION
**Pull request by mradigen**
This PR adds:
- Field coins to each team, defaulting to 0 on creation.
- Everytime a hint is requested, an amount is deducted from the team's coins. (amount can be set in `config.hint_penalty`)
- Team will not be allowed to view hints if they do not have the enough coins to do so.
- `/api/team/add` route creates a user and adds them to a given team. Returns 403 if user already exists in another team.
- `/api/team/remove` route removes a user from a team. Returns 403, if user is not a part of the selected team.

### CODE REVIEW

**Comment by mradigen** [comment]
@Meetesh-Saini How will we be incrementing the coins? Does solving the CTF reward the team with coins as well as leaderboard points?

**Comment by WizzyGeek** [comment]
> @Meetesh-Saini How will we be incrementing the coins? Does solving the CTF reward the team with coins as well as leaderboard points?

In the 1 week period, it reward them with just coins for solving challenges

**Comment by mradigen** [comment]
> > @Meetesh-Saini How will we be incrementing the coins? Does solving the CTF reward the team with coins as well as leaderboard points?
> 
> In the 1 week period, it rewards them with just coins for solving challenges

Got it. I'll implement an optional field called `coins` to `Problem`. The field can be set for the first week challenges, while the challenges on the event day would have a coin reward of 0.

**Review by WizzyGeek — APPROVED** [review]
the team's add and remove route needs to check the amount of Users in the team and other stuff, but I am not clear on the business logic for that functionality right now to specify the details myself

**Comment by WizzyGeek** [comment]
@mradigen get this merged asap, contact @Meetesh-Saini to merge it for you

File: src/pwncore/models/ctf.py [hunk 1]
```diff
@@ -22,13 +23,19 @@ class Problem(Model):
     points = fields.IntField()
     author = fields.TextField()
 
-    image_name = fields.TextField()
-    image_config: fields.Field[dict[str, list]] = fields.JSONField()  # type: ignore[assignment]
+    coins = fields.IntField(default=0)
```
Comment (WizzyGeek) [inline src/pwncore/models/ctf.py:26]: Inherit from Problems

**Review by WizzyGeek — CHANGES_REQUESTED** [review]

**Review by WizzyGeek — APPROVED** [review]

**Comment by mradigen** [comment]
The recent few commits add the following changes:
- Pre-event routes under `/api/ctf/pre/` which replicates `/api/ctf/`
- Splits event problem and pre-event problem models under `Problem` and `PreEventProblem` that inherit `BaseProblem`
- `Problem_Pydantic` has been changed to `BaseProblem_Pydantic`
- Normalizes User.tag using `.strip().casefold()`
- Coins under `PreEventProblem` is named as `points` (since it inherits from `BaseProblem`)
- Added `/api/leaderboard`

**Comment by WizzyGeek** [comment]
@mradigen ready to merge?

File: src/pwncore/routes/leaderboard.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+async def fetch_leaderboard():
+    teams = await Team.all()
+    leaderboard = {}
+    for team in teams:
+        problems_solved_by_team = await SolvedProblem.filter(
+            team__id=team.id
+        ).values_list("problem__points", flat=True)
+        # Tortoise's return type is List[Tuples] even after flat=True makes it List[]
+        # Hence ignore:
+        leaderboard[team.name] = sum(problems_solved_by_team)  # type: ignore[arg-type]
```
Comment (WizzyGeek) [inline src/pwncore/routes/leaderboard.py:22]: Use `Team.all().annotate(team_points = Sum("solvedproblems__problem__points"))`
adds a team_points field in the Result

Comment (WizzyGeek) [reply src/pwncore/routes/leaderboard.py:22]: image
Working fine

**Review by Meetesh-Saini — APPROVED** [review]

**Pull request merged**